### PR TITLE
Add spring cloud feature: manage azure spring cloud dependencies.

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+                    String springBootVer = springBootVersionMap.get("spring-boot-autoconfigure");
+                    if (StringUtils.isEmpty(springBootVer)) {
+                        throw new AzureExecutionException(String.format("Module %s is not a spring-boot application.", module.getName()));
+                    }
+                    ProgressManager.checkCanceled();
+
+                    progressIndicator.setText("Get latest versions ...");
+                    List<DependencyArtifact> dep = new ArrayList<>();
+                    dep.add(getDependencyArtifact(GROUP_ID, ARTIFACT_ID, versionMaps));
+                    ProgressManager.checkCanceled();
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java]: do we need to call `ProgressManager.checkCanceled();` after add one dependencyArtifact? If `getDependencyArtifact` cost to much time, shall we move the `ProgressManager.checkCanceled();` to the method?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyContextAction.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+package com.microsoft.azure.springcloud.dependency.action;
+
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.actionSystem.LangDataKeys;
+import com.intellij.openapi.actionSystem.Presentation;
+import com.intellij.openapi.module.Module;
+import org.jetbrains.annotations.NotNull;
+
+public class AddAzureDependencyContextAction extends AddAzureDependencyAction {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyContextAction.java]: when will we use this action?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/DependencyArtifact.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    public String getCurrentVersion() {
+        return currentVersion;
+    }
+
+    public void setCurrentVersion(String currentVersion) {
+        this.currentVersion = currentVersion;
+    }
+
+    private String groupId;
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/DependencyArtifact.java]: move the member variables to top?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyContextAction.java (hunk 2 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyContextAction.java]: AddAzureDependencyContextAction is used for the action triggered by right clicking at pom.xml

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java (hunk 1 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java]: Resolved getDependencyArtifact will not cost too much time

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/DependencyArtifact.java (hunk 3 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/DependencyArtifact.java]: Resolved in https://github.com/microsoft/azure-tools-for-java/pull/4211/commits/797f0796d6cec9897d9cee2c3cad8c240cf02fc1

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    private Document doc;
+
+    public SpringCloudDependencyManager(String effectivePomXml) throws DocumentException {
+        Map<String, String> nsContext = new HashMap<>();
+        nsContext.put("ns", POM_NAMESPACE);
+        DocumentFactory.getInstance().setXPathNamespaceURIs(nsContext);
+        doc = DocumentHelper.parseText(effectivePomXml);
+    }
+
+    public Map<String, String> getCurrentVersions(String groupId) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java]: getDependencyVersion?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        Map<String, String> res = new HashMap<>();
+        List<Node> nodes = doc.selectNodes("//ns:project/ns:dependencies/ns:dependency");
+        for (Node node : nodes) {
+            String thisGroupId = ((Element) node).elementTextTrim("groupId");
+            if (StringUtils.equals(thisGroupId, groupId)) {
+                res.put(((Element) node).elementTextTrim("artifactId"), ((Element) node).elementTextTrim("version"));
+            }
+        }
+
+        nodes = doc.selectNodes("//ns:project/ns:dependencyManagement/ns:dependencies/ns:dependency");
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java]: duplicate with previous codes, we could get all the nodes first and then get the dependency version

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        DocumentFactory.getInstance().setXPathNamespaceURIs(nsContext);
+        doc = DocumentHelper.parseText(effectivePomXml);
+    }
+
+    public Map<String, String> getCurrentVersions(String groupId) {
+        Map<String, String> res = new HashMap<>();
+        List<Node> nodes = doc.selectNodes("//ns:project/ns:dependencies/ns:dependency");
+        for (Node node : nodes) {
+            String thisGroupId = ((Element) node).elementTextTrim("groupId");
+            if (StringUtils.equals(thisGroupId, groupId)) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java]: thisGroupId -> dependencyGroupId
groupId -> targetGroupId

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        return res;
+    }
+
+    public static List<DependencyArtifact> getCompatibleVersion(List<DependencyArtifact> dependencies, String springBootVer)
+            throws AzureExecutionException, IOException, DocumentException {
+        List<DependencyArtifact> res = new ArrayList<>();
+        for (DependencyArtifact dependency : dependencies) {
+            List<String> latestVersions = getVersions(dependency.getGroupId(), dependency.getArtifactId());
+            if (latestVersions.isEmpty()) {
+                throw new AzureExecutionException((String.format("Cannot get version from maven central for: %s:%s.",
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java]: move the empty validation to `getVersions` method?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+                if (StringUtils.isEmpty(targetVer)) {
+                    targetVer = "2.2.0";
+                }
+                dependency.setCompilableVersion(targetVer);
+                if (!StringUtils.equals(dependency.getCurrentVersion(), targetVer)) {
+                    res.add(dependency);
+                }
+
+            } else if (springBootVer.startsWith("2.1.")) {
+                for (String ver2 : latestVersions) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java]: duplicated with previous code, shall we extract into one method?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            if (latestVersions.isEmpty()) {
+                throw new AzureExecutionException((String.format("Cannot get version from maven central for: %s:%s.",
+                        dependency.getGroupId(), dependency.getArtifactId())));
+            }
+            String targetVer = "";
+            if (springBootVer.startsWith("2.2.")) {
+
+                for (String ver2 : latestVersions) {
+                    if (ver2 != null && ver2.startsWith("2.2.")) {
+                        targetVer = ver2;
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/SpringCloudDependencyManager.java]: add break here? besides, shall we always use the latest ont?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/PomXmlUpdater.java [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+        // replace target node to placeholder
+        Element newNode = newNodes.get(0);
+        parent.content().replaceAll(t -> t == newNode ? placeHolderNode : t);
+        newNode.setParent(null);
+        // remove all spaces before target node
+        XmlUtils.trimTextBeforeEnd(parent, placeHolderNode);
+        final String xmlWithPlaceholder = parent.getDocument().asXML();
+
+        final List<String> newXmlLines = new ArrayList(
+                Arrays.asList(com.microsoft.azure.common.utils.TextUtils.splitLines(XmlUtils.prettyPrintElementNoNamespace(newNode))));
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/PomXmlUpdater.java]: no need to use full name here `com.microsoft.azure.common.utils.TextUtils`

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/PomXmlUpdater.java [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+            if (newNode != null) {
+                newNodes.add(newNode);
+            }
+        }
+        if (newNodes.isEmpty()) {
+            return;
+        }
+        if (dependenciesNode instanceof LocationAwareElement) {
+            for (int i = 1; i < newNodes.size(); i++) {
+                dependenciesNode.remove(newNodes.get(i));
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/PomXmlUpdater.java]: `dependenciesNode.subList`?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+                    List<DependencyArtifact> versionChanges = manager.getCompatibleVersion(dep, springBootVer);
+                    if (versionChanges.isEmpty()) {
+                        PluginUtil.showInfoNotificationProject(project, "Your project is update-to-date.",
+                                "No updates are needed.");
+                        return;
+                    }
+                    ProgressManager.checkCanceled();
+                    progressIndicator.setText("Applying versions ...");
+                    File pomFile = new File(mavenProject.getFile().getCanonicalPath());
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java]: Do we support multi pom file?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/PomXmlUpdater.java (hunk 11 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/PomXmlUpdater.java]: No such method

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java (hunk 12 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/springcloud/dependency/action/AddAzureDependencyAction.java]: No at current stage

**Pull request merged**